File: scala/raindrops/src/main/scala/Raindrops.scala
object Raindrops {
  private[this] val DROP_FACTORS = Map(
    3 -> "Pling",
    5 -> "Plang",
    7 -> "Plong"
  )

  def convert(n: Int): String = {
    val sentence = DROP_FACTORS.filter({
      case (k, _) => n % k == 0
    }).map({
      case (_, sentence) => sentence
    }).mkString("")
    if (sentence.isEmpty)
      n.toString
    else
      sentence
  }
}

File: haskell/raindrops/src/Raindrops.hs
module Raindrops (convert) where

import Data.Maybe

convert :: Int -> String
convert n = if drops == "" then show n else drops
    where
        dict = [(3, "Pling"), (5, "Plang"), (7, "Plong")]
        drops = concat $ catMaybes [if n `mod` (fst r) == 0 then Just $ snd r else Nothing | r <- dict]
    
            